File: 1960-math-cake.scala
//> using scala 3.7.1

package olympus.mathcake

import scala.annotation.tailrec

@main
def main(cakePieces: Int, people: Int): Unit =
  if cakePieces < people then println("not enough cake pieces")
  else if people == 1 || cakePieces == people then println("1")
  else
    val surplus = cakePieces - people
    val distribution = Array.fill(people)(1)
    distribution(people - 1) = surplus + 1

    val initDist = distribution.toVector
    println(countAllDists(initDist, None))

@tailrec
def countAllDists(
    initDist: Vector[Int],
    prevReceiverIndex: Option[Int],
    currentCount: Int = 1
): Int =
  println(initDist)
  nextDistribution(initDist, prevReceiverIndex) match
    case None => currentCount
    case Some((newDist, ri, di)) =>
      tryFastForwardDistribution(newDist, ri, di) match
        case None => countAllDists(newDist, Some(ri), currentCount + 1)
        case Some((futureDist, iterations)) =>
          countAllDists(futureDist, Some(ri), currentCount + 1 + iterations)

def tryFastForwardDistribution(
    dist: Vector[Int],
    rIndex: Int,
    dIndex: Int
): Option[(Vector[Int], Int)] =
  if rIndex == dIndex - 1 then
    val receiver = dist(rIndex)
    val donor = dist(dIndex)
    val diff = donor - receiver
    if diff > 1 then
      val iterations: Int = diff / 2
      val futureDist = dist.toArray
      futureDist(rIndex) = receiver + iterations
      futureDist(dIndex) = donor - iterations

      println(dist)
      println("fast-forwarding...")
      println(s"calculated iterations: $iterations")

      Some((futureDist.toVector, iterations))
    else None
  else None

def nextDistribution(
    dist: Vector[Int],
    prevReceiverIndex: Option[Int]
): Option[(Vector[Int], Int, Int)] =
  for (ri, di) <- findReceiverDonorIndices(dist)
  yield
    val receiver = dist(ri)
    val donor = dist(di)
    val newDist = dist.toArray
    newDist(ri) = receiver + 1
    newDist(di) = donor - 1

    val prevRi = prevReceiverIndex.getOrElse(ri)
    if ri < prevRi then
      println("rebalancing")
      // rebalance distribution - shift all extra diffs to the last element
      val baseLine = receiver + 1
      val lastIndex = newDist.length - 1

      for i <- (ri + 1) until lastIndex do
        val el = newDist(i)
        if el > baseLine then
          val diff = el - baseLine
          newDist(i) = baseLine
          newDist(lastIndex) += diff

    (newDist.toVector, ri, di)

def findReceiverDonorIndices(dist: Vector[Int]): Option[(Int, Int)] =
  var index = dist.length - 2
  findReceiverDonorIndices(dist, index)

@tailrec
def findReceiverDonorIndices(
    dist: Vector[Int],
    rIndex: Int
): Option[(Int, Int)] =
  if rIndex < 0 then None
  else if dist(rIndex) < dist(rIndex + 1) then
    findDonorIndex(dist, rIndex) match
      case Some(donorIndex) => Some((rIndex, donorIndex))
      case None             => findReceiverDonorIndices(dist, rIndex - 1)
  else findReceiverDonorIndices(dist, rIndex - 1)

def findDonorIndex(dist: Vector[Int], receiverIndex: Int): Option[Int] =
  val dIndex = receiverIndex + 1
  findDonorIndex(dist, receiverIndex, dIndex)

@tailrec
def findDonorIndex(
    dist: Vector[Int],
    receiverIndex: Int,
    dIndex: Int
): Option[Int] =
  if dIndex >= dist.length then None
  else
    val receiver = dist(receiverIndex)
    val donor = dist(dIndex)
    val beforeDonor = dist(dIndex - 1)
    if donor > receiver + 1 && donor > beforeDonor then Some(dIndex)
    else findDonorIndex(dist, receiverIndex, dIndex + 1)

File: 1960-math-cake.py
import sys
from itertools import repeat
from typing import Optional

def main(cake_pieces: int, people: int) -> None:
  if cake_pieces < people:
    print("not enough cake pieces")
  elif people == 1 or cake_pieces == people:
    print("1")
  else:
    surplus = cake_pieces - people
    distribution = list(repeat(1, people))
    distribution[people - 1] = surplus + 1

    print(countAllDists(distribution))


def countAllDists(
  dist: list[int], prev_receiver_index: int = 0, current_count: int = 1
) -> int:
  while True:
    print(dist)
    next_d = next_distribution(dist, prev_receiver_index)
    if next_d is None:
      return current_count
    else:
      dist, ri, di = next_d
      prev_receiver_index = ri

      current_count += 1

      ffd = try_fast_forward_distribution(dist, ri, di)
      if not ffd is None:
        dist, iterations = ffd
        current_count += iterations

def try_fast_forward_distribution(
  dist: list[int], r_index: int, d_index: int
) -> Optional[tuple[list[int], int]]:
  if r_index == d_index - 1:
    receiver = dist[r_index]
    donor = dist[d_index]
    diff = donor - receiver
    if diff > 1:
      print(dist)
      print("fast-forwarding...")

      iterations = diff // 2
      dist[r_index] = receiver + iterations
      dist[d_index] = donor - iterations

      print("calculated iterations:", iterations)

      return (dist, iterations)
    else:
      return None
  else:
    return None

def next_distribution(
  dist: list[int], prev_receiver_index: int = 0
) -> Optional[tuple[list[int], int, int]]:
  indices = find_receiver_donor_indices(dist, len(dist) - 2)
  if indices is None:
    return None
  else:
    ri, di = indices
    receiver = dist[ri]
    donor = dist[di]
    dist[ri] = receiver + 1
    dist[di] = donor - 1

    if ri < prev_receiver_index:
      print("rebalancing")
      # rebalance distribution - shift all extra diffs to the last element
      base_line = receiver + 1
      last_index = len(dist) - 1

      for i in range(ri + 1, last_index):
        el = dist[i]
        if el > base_line:
          diff = el - base_line
          dist[i] = base_line
          dist[last_index] += diff

    return (dist, ri, di)

def find_receiver_donor_indices(dist: list[int], r_index: int) -> Optional[tuple[int, int]]:
  if r_index < 0:
    return None
  elif dist[r_index] < dist[r_index + 1]:
    donor_index = find_donor_index(dist, r_index, r_index + 1)
    if not donor_index is None:
      return (r_index, donor_index)
    else:
      return find_receiver_donor_indices(dist, r_index - 1)
  else:
    return find_receiver_donor_indices(dist, r_index - 1)

def find_donor_index(dist: list[int], receiver_index: int, d_index: int) -> Optional[int]:
  if d_index >= len(dist):
   return None
  else:
    receiver = dist[receiver_index]
    donor = dist[d_index]
    beforeDonor = dist[d_index - 1]
    if donor > receiver + 1 and donor > beforeDonor:
      return d_index
    else:
      return find_donor_index(dist, receiver_index, d_index + 1)

if __name__ == "__main__":
  args = sys.argv
  if len(args) > 2:
    cake_pieces = int(args[1])
    people = int(args[2])
  else:
    cake_pieces = int(input())
    people = int(input())

  main(cake_pieces, people)
